---
title: "Demultiplex report"
date: "`r format(Sys.time(), '%d %B, %Y')`"
params:
   rmd: "report_demultiplex.Rmd"
output: html_document
editor_options: 
  chunk_output_type: console
---

<style>
.main-container {
    width: 70%;
    max-width: unset;
}
</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
if (!requireNamespace("DT", quietly = TRUE)) install.packages('DT', repos = "https://cloud.r-project.org")
if (!requireNamespace("ggplot2", quietly = TRUE)) install.packages('ggplot2', repos = "https://cloud.r-project.org")
if (!requireNamespace("dplyr", quietly = TRUE)) install.packages('dplyr', repos = "https://cloud.r-project.org")
if (!requireNamespace("stringr", quietly = TRUE)) install.packages('stringr', repos = "https://cloud.r-project.org")
if (!requireNamespace("S4Vectors", quietly = TRUE)) install.packages('S4Vectors', repos = "https://cloud.r-project.org")
if (!requireNamespace("Biostrings", quietly = TRUE)) {
  if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager", repos = "https://cloud.r-project.org")
  BiocManager::install("Biostrings", update = F, ask = F)
}
if (!requireNamespace("qckitfastq", quietly = TRUE)) {
  if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager", repos = "https://cloud.r-project.org")
  BiocManager::install("qckitfastq")
}

library(stringr)
library(ggplot2)
library(dplyr)
library(DT)
library(Biostrings)
library(S4Vectors)
library(parallel)

# library(BiocParallel)
# register(MulticoreParam(workers = ncores, progressbar = TRUE), default = TRUE)
# register(SnowParam(workers = ncores, progressbar = FALSE))
# registered()

get_from_fq <- function(fg, step) {
  fastq <- Biostrings::readDNAStringSet(fg, format = "fastq", seek.first.rec = T, with.qualities = T)
  #fastq <- Biostrings::readDNAStringSet("/data3/projects/xanthopore/seb/bd_output/reads_per_genome/fastq/CIX3092.fastq.gz", format = "fastq", seek.first.rec = T, with.qualities = T)

  if (length(fastq) == 0) {
    warning(fg , " is empty!")
    return(data.frame())
  }
  cat("#####  Processing: ", fg, "\n")
  seqNames <- names(fastq)
  seqWidth <- width(fastq)
  phred <- Biostrings::PhredQuality(mcols(fastq)$qualities)
  rm(fastq)
  scores <- sapply(as(phred, "IntegerList"), mean)
  dt <- data.frame("read_id" = gsub(" .+", "", seqNames),
                   "run_id" = gsub("(.+ runid=| sampleid.+)", "", seqNames),
                   "read_length" = seqWidth,
                   "phred.score" = scores,
                   "Genome_ID" = gsub("\\.fastq.gz", "", basename(fg)),
                   "Step" = step,
                   "id" = seqNames,
                   stringsAsFactors = F)
  return(dt)
}


get_from_fq_lw <- function(fg, step) {
  message("#####  Processing: ", fg, "\n")
  fastq <- Biostrings::readDNAStringSet(fg, format = "fastq", seek.first.rec = T, with.qualities = FALSE)
  if (length(fastq) == 0) {
    warning(fg , " is empty!")
    return(data.frame())
  }
  seqNames <- names(fastq)
  seqWidth <- width(fastq)
  rm(fastq)
  pscores <- qckitfastq::per_read_quality(fg, output_file = "plot")$sequence_mean
  dt <- data.frame("read_id" = gsub(" .+", "", seqNames),
                   "run_id" = gsub("(.+ runid=| sampleid.+)", "", seqNames),
                   "read_length" = seqWidth,
                   "phred.score" = pscores,
                   "Genome_ID" = gsub("\\.fastq.gz", "", basename(fg)),
                   "Step" = step,
                   "id" = seqNames,
                   stringsAsFactors = F)
  return(dt)
}

find_n50 <- function(x) {
  thredshold <- sum(x)/2
  test <- 0
  for (l in sort(x)) {
    test <- test + l
    if (test > thredshold) break
  }
  return(l)
}
```


```{r getting parameters from snakemake object, eval = T}
ncores <- as.numeric(snakemake@threads)

fastq_folder <- snakemake@params$fastq # "/data3/projects/xanthopore/baotram/test/results_guppy/reads_per_genome/fastq"

demux_folder <- snakemake@params$demultiplex # "/data3/projects/xanthopore/baotram/test/results_guppy/demultiplex"

postdemux_folder <- snakemake@params$postdemux # "/data3/projects/xanthopore/baotram/test/results_guppy/reads_per_genome/fastq_porechop_filtlong"

bcgenome_tsv <- snakemake@params$barcode_by_genome # "/data3/projects/xanthopore/baotram/baseDmux/barcodeByGenome_sample.tsv"
```


```{r getting parameters for tests, eval = F}
fastq_folder <- "/home/cunnac/TEMP/test_baseDmux/result/reads_per_genome/fastq"
demux_folder <- "/home/cunnac/TEMP/test_baseDmux/result/demultiplex"
postdemux_folder <- "/home/cunnac/TEMP/test_baseDmux/result/reads_per_genome/fastq_porechop_filtlong1"
bcgenome_tsv <- "/home/cunnac/TEMP/test_baseDmux/configFiles_local/profile/barcodeByGenome_gupdeep.tsv"
ncores <- 1
```

# Menu {.tabset .tabset-fade}


## session info {.tabset .tabset-fade}

This report provides views on aggregated read metrics, considering only reads passing the specified basecalling quality cutoff. These metrics are contrasted by a number of factors including the ONT run instance, assigned barcode by the demultiplexer program (guppy and deepbinner), ...  

Note that the graphs in the `sequencing_summary` tab are constructed using data obtained directly from sequencing_summary.txt files.  
  
In contrast, graphs in other tabs are constructed with data recomputed using reads in genome/sample bins fastq files. In addition, if those reads were further processed with porechop and/or filtlong, the corresponding data is displayed. Obviously, these graphs only consider reads that could be assigned as belonging to a specific barcode and thus, a genome.  


```{r R session info}
sessionInfo()
```


## sequencing_summary {.tabset .tabset-fade}

```{r get demultiplexer and barcode specific sequencing_summary files info}
summaries <- list.files(demux_folder, "sequencing_summary.txt", recursive = T, full.names = T)
sum <- bind_rows(
               mclapply(summaries,
                        function(s) {
                          sum <- read.table(s, header = T)
                          splitted_path <- unlist(strsplit(dirname(s), "/", fixed = T))
                          if (nrow(sum) > 0) {
                            sum <- data.frame(sum,
                                              "ONT_Barcode" = splitted_path[length(splitted_path)],
                                              "Run_ID" = splitted_path[length(splitted_path) - 1],
                                              "Demultiplexer" = splitted_path[length(splitted_path) - 2],
                                              stringsAsFactors = F)
                            }
                          return(sum)
                          }, mc.cores = ncores
                        )
               ) %>% as_tibble()

```


```{r aggregating seq sum files data}
seqSumAggregData <- sum %>%
  dplyr::group_by(Demultiplexer, Run_ID, ONT_Barcode) %>%
  dplyr::summarise(total.gigabases = sum(sequence_length_template)/10**9,
                   total.reads = length(sequence_length_template),
                   mean.length = mean(sequence_length_template),
                   median.length = median(sequence_length_template),
                   N50.length = find_n50(sequence_length_template),
                   mean.quality.score = mean(mean_qscore_template),
                   median.quality.score = median(mean_qscore_template),
                   N50.quality.score = find_n50(mean_qscore_template))
```



### Total gigabases

```{r seqsum.total.gigabases, fig.width = 16, fig.height = 8}
ggplot(seqSumAggregData, aes(x = ONT_Barcode, y = total.gigabases, fill = Demultiplexer)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0), stat = "identity") +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Total reads

```{r seqsum.total.reads, fig.width = 16, fig.height = 8}
ggplot(seqSumAggregData, aes(x = ONT_Barcode, y = total.reads, fill = Demultiplexer)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol=3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Mean length

```{r seqsum.mean.length, fig.width = 16, fig.height = 8}
ggplot(seqSumAggregData, aes(x = ONT_Barcode, y = mean.length, fill = Demultiplexer)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Median length

```{r seqsum.median.length, fig.width = 16, fig.height = 8}
ggplot(seqSumAggregData, aes(x = ONT_Barcode, y = median.length, fill = Demultiplexer)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x=element_text(angle=90,hjust=1),
        legend.direction = "horizontal", legend.position = "top")
```

### N50 length

```{r seqsum.N50.length, fig.width = 16, fig.height = 8}
ggplot(seqSumAggregData, aes(x = ONT_Barcode, y = N50.length, fill = Demultiplexer)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Mean quality score

```{r seqsum.mean.quality, fig.width = 16, fig.height = 8}
ggplot(seqSumAggregData, aes(x = ONT_Barcode, y = mean.quality.score, fill = Demultiplexer)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Median quality score

```{r seqsum.median.quality, fig.width = 16, fig.height = 8}
ggplot(seqSumAggregData, aes(x = ONT_Barcode, y = median.quality.score, fill = Demultiplexer)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```


### Quality score distribution

```{r seqsum.quality.distribution, fig.width = 16, fig.height = 8}
ggplot(sum, aes(x = round(mean_qscore_template))) +
  geom_line(aes(color = ONT_Barcode), stat = "count") +
  facet_grid(cols = vars(Run_ID), rows = vars(Demultiplexer), scales = "fixed") +
  scale_color_viridis_d(option = "E", end = .85, alpha = 0.8) +
  theme_light() +
  theme(axis.text.x=element_text(angle=90,hjust=1),
        legend.direction = "horizontal", legend.position = "top")
```


```{r export seqsum data, include=FALSE, eval = T}
image_path <- file.path(snakemake@params$outpath, "demultiplex_report.RData")
save.image(file = image_path, compress = TRUE)
```



## fqs by run {.tabset .tabset-fade}


```{r loading barcodeByGenome table, message=TRUE}
if (!file.exists(bcgenome_tsv)) {
  message("Cannot find the provided roadmap barcodeByGenome file at this location:\n", bcgenome_tsv)
  message("This file is needed to proceed... Exiting prematurely!!!\n")
  knitr::knit_exit()
}
roadmap <- read.delim(bcgenome_tsv)
```



```{r computing metrics from reads in genome bins fastq files}
demux_fastq <- list.files(fastq_folder, ".*\\.fastq.gz", full.names = T)
demux_fastq <- demux_fastq[grep("porechop|filtlong", demux_fastq, invert = T)]

demux <- bind_rows(
  mclapply(X = demux_fastq, FUN = get_from_fq_lw, step = "demultiplex",
           mc.preschedule = FALSE, mc.cores = ncores)
  )

demux <- demux %>% 
  # figuring out the demultiplexer associated with a given genome
  left_join(roadmap %>% select(Demultiplexer, Genome_ID) %>% distinct(),
            by = c("Genome_ID" = "Genome_ID"))

```




```{r include barcode id in read metrics table}
demux_sum <- dplyr::inner_join(demux,
                              dplyr::select(sum, read_id, run_id, ONT_Barcode, Run_ID, Demultiplexer)
)
```



```{r postdemux}
if (any(postdemux_folder != demux_folder)) {
  postdemux_fastq <- list.files(postdemux_folder, ".*\\.fastq.gz", full.names = T)
  postdemux_fastq <- postdemux_fastq[grep("porechop|filtlong", postdemux_fastq)]

  # postdemux <- do.call(rbind, mclapply(postdemux_fastq, function(fg) {
  #   fastq <- readDNAStringSet(fg, format = "fastq", seek.first.rec = T, with.qualities = T)
  #   # phred score
  #   phred <- PhredQuality(mcols(fastq)$qualities)
  #   scores <- unlist(lapply(as(phred, "IntegerList"), mean))
  #
  #   dt <- data.frame("read_id" = gsub(" .+", "", names(fastq)),
  #                    "run_id" = gsub("(.+ runid=| sampleid.+)", "", names(fastq)),
  #                    "read_length" = width(fastq),
  #                    "phred.score" = scores,
  #                    # "Genome_ID" = gsub("\\.fastq.gz", "", basename(fg)),
  #                    "Step" = gsub("fastq", "demultiplex", basename(dirname(fg))),
  #                    "id" = names(fastq),
  #                    stringsAsFactors = F)
  #   return(dt)
  # }))

  postdemux <-  bind_rows(mcmapply(FUN = get_from_fq_lw,
                            fg = postdemux_fastq,
                            step = gsub("fastq", "demultiplex", basename(dirname(postdemux_fastq))),
                            SIMPLIFY = FALSE,
                            mc.preschedule = FALSE, mc.cores = ncores)
                   )
#  postdemux$Genome_ID <- NULL

  postdemux <- postdemux %>% mutate(Genome_ID = gsub(gsub("fastq", "", basename(postdemux_folder)),
                                        "",
                                        Genome_ID)
                       )
  
  postdemux <- postdemux %>% 
  left_join(roadmap %>% select(Demultiplexer, Genome_ID) %>% distinct(),
            by = c("Genome_ID" = "Genome_ID"))
  
  postdemux_sum <- dplyr::inner_join(postdemux,
                              dplyr::select(sum, read_id, run_id, ONT_Barcode, Run_ID, Demultiplexer)
                              )

#  postdemux <- merge(postdemux, demux[c("read_id", "run_id", "Genome_ID")], by = c("read_id", "run_id"))
  # postdemux_sum <- merge(postdemux, sum[c("read_id", "run_id", "ONT_Barcode", "Run_ID", "Demultiplexer")], by = c("read_id", "run_id"))
}
```



```{r all_qc}
if (any(postdemux_folder == demux_folder)) {
  all_sum <- demux_sum %>% dplyr::mutate(Step = NA)
} else {
  all_sum <- rbind(demux_sum, postdemux_sum)
}

all_sum$Demultiplexer <- factor(all_sum$Demultiplexer, levels = c("guppy", "deepbinner"), ordered = F)

byrun <- all_sum %>%
  # dplyr::filter(Step == "demultiplex") %>%
  dplyr::group_by(Step, Demultiplexer, Run_ID, ONT_Barcode, Genome_ID) %>%
  dplyr::summarise(total.gigabases = sum(read_length)/10**9,
                   total.reads = length(read_length),
                   mean.length = mean(read_length),
                   median.length = median(read_length),
                   N50.length = find_n50(read_length),
                   mean.quality.score = mean(phred.score),
                   median.quality.score = median(phred.score),
                   N50.quality.score = find_n50(phred.score))

```


```{r, eval = T}
if (nlevels(factor(byrun$Step)) > 1) {
  cat("FILTERING PARAMS\n\n")
  if ("porechop" %in% names(snakemake@config)) print(snakemake@config["porechop"])
  filtlongs <- names(snakemake@config)[grep("filtlong.+", names(snakemake@config))]
  if (length(filtlongs) > 0) print(snakemake@config[filtlongs])
}
```

### Total gigabases

```{r run.total.gigabases, fig.width = 16, fig.height = 8}
ggplot(byrun, aes(x = ONT_Barcode, y = total.gigabases, fill = Demultiplexer, alpha = Step)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0), stat = "identity") +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Total reads

```{r run.total.reads, fig.width = 16, fig.height = 8}
ggplot(byrun, aes(x = ONT_Barcode, y = total.reads, fill = Demultiplexer, alpha = Step)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol=3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Mean length

```{r run.mean.length, fig.width = 16, fig.height = 8}
ggplot(byrun, aes(x = ONT_Barcode, y = mean.length, fill = Demultiplexer, alpha = Step)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Median length

```{r run.median.length, fig.width = 16, fig.height = 8}
ggplot(byrun, aes(x = ONT_Barcode, y = median.length, fill = Demultiplexer, alpha = Step)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x=element_text(angle=90,hjust=1),
        legend.direction = "horizontal", legend.position = "top")
```

### N50 length

```{r run.N50.length, fig.width = 16, fig.height = 8}
ggplot(byrun, aes(x = ONT_Barcode, y = N50.length, fill = Demultiplexer, alpha = Step)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Mean quality score

```{r run.mean.quality, fig.width = 16, fig.height = 8}
ggplot(byrun, aes(x = ONT_Barcode, y = mean.quality.score, fill = Demultiplexer, alpha = Step)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### Median quality score

```{r run.median.quality, fig.width = 16, fig.height = 8}
ggplot(byrun, aes(x = ONT_Barcode, y = median.quality.score, fill = Demultiplexer, alpha = Step)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```

### N50 quality score

```{r run.n50.quality, fig.width = 16, fig.height = 8}
ggplot(byrun, aes(x = ONT_Barcode, y = N50.quality.score, fill = Demultiplexer, alpha = Step)) +
  geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
  scale_fill_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  facet_wrap(vars(Run_ID), scales = "fixed", ncol = 3, strip.position = "right") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        legend.direction = "horizontal", legend.position = "top")
```


### Quality score distribution

```{r run.quality.distribution, fig.width = 16, fig.height = 8}
ggplot(all_sum, aes(x = round(phred.score))) +
  geom_line(aes(alpha = Step, color = Demultiplexer), stat = "count") +
  facet_grid(cols = vars(Run_ID), rows = vars(Demultiplexer), scales = "fixed") +
  scale_color_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  theme_light() +
  theme(axis.text.x=element_text(angle=90,hjust=1),
        legend.direction = "horizontal", legend.position = "top")
```

## fqs by genome {.tabset .tabset-fade}

```{r}
if (file.exists(bcgenome_tsv)) {
  bygenome <- byrun
}
```

```{r, eval = T}
if (file.exists(bcgenome_tsv)) {
  if (nlevels(factor(bygenome$Step)) > 1) {
  cat("FILTERING PARAMS\n\n")
  if ("porechop" %in% names(snakemake@config)) print(snakemake@config["porechop"])
  filtlongs <- names(snakemake@config)[grep("filtlong.+", names(snakemake@config))]
  if (length(filtlongs) > 0) print(snakemake@config[filtlongs])
  }
}
```

### Total gigabases

```{r genome.total.gigabases, fig.width = 16, fig.height = 10}
if (file.exists(bcgenome_tsv)) {
  p <- ggplot(bygenome, aes(x = Genome_ID, y = total.gigabases, fill = Run_ID, color = ONT_Barcode)) +
    geom_col(position = position_stack()) +
    geom_text(aes(label = gsub("[a-z]", "", ONT_Barcode)), color = "#b1bed5", position = position_stack(), vjust = 1.2, size = 3.88*16/nlevels(as.factor(bygenome$Genome_ID))) +
    scale_color_brewer(palette = "Paired", guide = NULL) +
    scale_fill_viridis_d(option = "C") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.direction = "vertical", legend.position = "top")

if (!all(is.na(bygenome$Step))) p <- p + facet_wrap(vars(Step), nrow = nlevels(factor(bygenome$Step)))
  p
}
```




### Total reads

```{r genome.total.reads, fig.width = 16, fig.height = 10}
if (file.exists(bcgenome_tsv)) {
  p <- ggplot(bygenome, aes(x = Genome_ID, y = total.reads, fill=Run_ID)) +
    geom_col(position = position_stack(), color = "gray") +
    geom_text(aes(label = gsub("[a-z]", "", ONT_Barcode)), color = "#b1bed5", position = position_stack(), vjust = 1.2, size = 3.88*16/nlevels(as.factor(bygenome$Genome_ID))) +
    scale_fill_viridis_d(option = "C") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.direction = "vertical", legend.position = "top")

  if (!all(is.na(bygenome$Step))) p <- p + facet_wrap(vars(Step), nrow = nlevels(factor(bygenome$Step)))
  p
}
```

### Mean length

```{r genome.mean.length, fig.width = 16, fig.height = 10}
if (file.exists(bcgenome_tsv)) {
  p <- ggplot(bygenome, aes(x = Genome_ID, y = mean.length, fill = Run_ID)) +
    geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
    geom_text(aes(label = gsub("[a-z]", "", ONT_Barcode)), color = "#b1bed5", position = position_dodge2(width = 1, preserve = "single", padding = 0), angle = 90, vjust= .5, hjust = 1, size = 3.88*2/nlevels(as.factor(bygenome$Step))) +
    scale_fill_viridis_d(option = "C") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.direction = "vertical", legend.position = "top")

  if (!all(is.na(bygenome$Step))) p <- p + facet_wrap(vars(Step), nrow = nlevels(factor(bygenome$Step)))
  p
}
```

### Median length

```{r genome.median.length, fig.width = 16, fig.height = 10}
if (file.exists(bcgenome_tsv)) {
  p <- ggplot(bygenome, aes(x = Genome_ID, y = median.length, fill = Run_ID)) +
    geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
    geom_text(aes(label = gsub("[a-z]", "", ONT_Barcode)), color = "#b1bed5", position = position_dodge2(width = 1, preserve = "single", padding = 0), angle = 90, vjust= .5, hjust = 1, size = 3.88*2/nlevels(as.factor(bygenome$Step))) +
    scale_fill_viridis_d(option = "C") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.direction = "vertical", legend.position = "top")

  if (!all(is.na(bygenome$Step))) p <- p + facet_wrap(vars(Step), nrow = nlevels(factor(bygenome$Step)))
  p
}
```

### N50 length

```{r genome.N50.length, fig.width = 16, fig.height = 10}
if (file.exists(bcgenome_tsv)) {
  p <- ggplot(bygenome, aes(x = Genome_ID, y = N50.length, fill = Run_ID)) +
    geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
    geom_text(aes(label = gsub("[a-z]", "", ONT_Barcode)), color = "#b1bed5", position = position_dodge2(width = 1, preserve = "single", padding = 0), angle = 90, vjust= .5, hjust = 1, size = 3.88*2/nlevels(as.factor(bygenome$Step))) +
    scale_fill_viridis_d(option = "C") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.direction = "vertical", legend.position = "top")

  if (!all(is.na(bygenome$Step))) p <- p + facet_wrap(vars(Step), nrow = nlevels(factor(bygenome$Step)))
  p
}
```

### Mean quality score

```{r genome.mean.quality, fig.width = 16, fig.height = 10}
if (file.exists(bcgenome_tsv)) {
  p <- ggplot(bygenome, aes(x = Genome_ID, y = mean.quality.score, fill = Run_ID)) +
    geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
    geom_text(aes(label = gsub("[a-z]", "", ONT_Barcode)), color = "#b1bed5", position = position_dodge2(width = 1, preserve = "single", padding = 0), angle = 90, vjust= .5, hjust = 1, size = 3.88*2/nlevels(as.factor(bygenome$Step))) +
    scale_fill_viridis_d(option = "C") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.direction = "vertical", legend.position = "top")

  if (!all(is.na(bygenome$Step))) p <- p + facet_wrap(vars(Step), nrow = nlevels(factor(bygenome$Step)))
  p
}
```

### Median quality score

```{r genome.median.quality, fig.width = 16, fig.height = 10}
if (file.exists(bcgenome_tsv)) {
  p <- ggplot(bygenome, aes(x = Genome_ID, y = median.quality.score, fill = Run_ID)) +
    geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
    geom_text(aes(label = gsub("[a-z]", "", ONT_Barcode)), color = "#b1bed5", position = position_dodge2(width = 1, preserve = "single", padding = 0), angle = 90, vjust= .5, hjust = 1, size = 3.88*2/nlevels(as.factor(bygenome$Step))) +
    scale_fill_viridis_d(option = "C") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.direction = "vertical", legend.position = "top")

  if (!all(is.na(bygenome$Step))) p <- p + facet_wrap(vars(Step), nrow = nlevels(factor(bygenome$Step)))
  p
}
```

### N50 quality score

```{r genome.n50.quality, fig.width = 16, fig.height = 10}
if (file.exists(bcgenome_tsv)) {
  p <- ggplot(bygenome, aes(x = Genome_ID, y = N50.quality.score, fill = Run_ID)) +
    geom_col(position = position_dodge2(preserve = "single", padding = 0)) +
    geom_text(aes(label = gsub("[a-z]", "", ONT_Barcode)), color = "#b1bed5", position = position_dodge2(width = 1, preserve = "single", padding = 0), angle = 90, vjust= .5, hjust = 1, size = 3.88*2/nlevels(as.factor(bygenome$Step))) +
    scale_fill_viridis_d(option = "C") +
    theme_light() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1),
          legend.direction = "vertical", legend.position = "top")

  if (!all(is.na(bygenome$Step))) p <- p + facet_wrap(vars(Step), nrow = nlevels(factor(bygenome$Step)))
  p
}
```

### Quality score distribution

```{r genome.quality.distribution, fig.width = 16, fig.height = 10}
if (file.exists(bcgenome_tsv)) {
  ggplot(all_sum, aes(x = round(phred.score))) +  
  geom_line(aes(color = Demultiplexer), stat = "count") +
  facet_grid(cols = vars(Genome_ID), rows = vars(Step), scales = "fixed") +
  scale_color_viridis_d(option = "E", end = .85) +
  scale_alpha_discrete(range = c(.3, 1)) +
  theme_light() +
  theme(axis.text.x=element_text(angle=90,hjust=1),
        legend.direction = "horizontal", legend.position = "top")
}
```




## data

```{r barcode_table}
if (file.exists(bcgenome_tsv)) {
  datatable(bygenome)
} else {
  datatable(byrun)
}
```

```{r export all data, include=FALSE, eval = T}
if (file.exists(bcgenome_tsv)) {
  write.table(byrun, file.path(snakemake@params$outpath, "demultiplex_report_fastqs.tsv"), sep = "\t", row.names = F, quote = F)
}
image_path <- file.path(snakemake@params$outpath, "demultiplex_report.RData")
remove(p)
save.image(file = image_path, compress = TRUE)
```
